### PR TITLE
Color and alpha blending bug fix for scenes with transparent gl3d objects

### PR DESCRIPTION
**Pull request by archmoj**
Addressing an old bug in `gl3d` scenes with transparent objects affected `scatter3d`, `mesh3d`, `surface`, `volume` and possibly other 3-D traces:
https://github.com/plotly/plotly.js/issues/1267
https://github.com/plotly/plotly.js/issues/3243
https://github.com/plotly/plotly.js/issues/4111

[One codepen for comparing before vs after of all 3d mocks with transparency added](https://codepen.io/MojtabaSamimi/pen/bGbPBGa?editors=0010)

[Before](https://codepen.io/MojtabaSamimi/pen/jONoJqE?editors=0010)
[After](https://codepen.io/MojtabaSamimi/pen/RwbmdrY?editors=0010)

[Before](https://codepen.io/MojtabaSamimi/pen/KOQmOK?editors=0010)
[After](https://codepen.io/MojtabaSamimi/pen/gOYJqVB?editors=1000)

[Before](https://codepen.io/etpinard/pen/oYOMZO)
[After](https://codepen.io/MojtabaSamimi/pen/PoYvLYX)

[Before](https://codepen.io/MojtabaSamimi/pen/LYPKpxQ?editors=0010)
[After](https://codepen.io/MojtabaSamimi/pen/WNeqQoe?editors=0010)

As well as point clouds (`scatter3d` plots), this PR improves renderings of volumes and surfaces.

Also please refer to https://github.com/plotly/plotly.js/pull/4234/commits/3632eea636ce23a3a4adf5a336daee5ef5edfad4 that illustrates some issues fixed in this PR.

cc: @plotly/plotly_js

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -81,7 +81,7 @@
     "gl-mat4": "^1.2.0",
     "gl-mesh3d": "^2.1.1",
     "gl-plot2d": "^1.4.2",
-    "gl-plot3d": "^2.2.2",
+    "gl-plot3d": "git://github.com/gl-vis/gl-plot3d.git#67b5aed84d02429bbc4e8c5a45d4cd23e03280c6",
```
Comment (archmoj) [inline package.json]: https://github.com/gl-vis/gl-plot3d/pull/22

File: test/image/mocks/gl3d_volume_multiple-traces.json [hunk 2]
```diff
@@ -6,10 +6,10 @@
             "lighting": {
                 "facenormalsepsilon": 0
             },
-            "opacity": 0.2,
```
Comment (archmoj) [inline test/image/mocks/gl3d_volume_multiple-traces.json]: This opacity for a `volume` trace was too high.

File: test/image/mocks/gl3d_volume_multiple-traces.json [hunk 3]
```diff
@@ -6,10 +6,10 @@
...
+            "opacity": 0.1,
             "opacityscale": "extremes",
-            "colorscale": "Blackbody",
```
Comment (archmoj) [inline test/image/mocks/gl3d_volume_multiple-traces.json]: This was a bad `colorscale` choice for a volume with *extreme* `opacityscale`.

**Comment by etpinard** [comment]
Hmm. At first look, the new baseline for the `gl3d_opacity-scaling-spikes` mock looks worse than before:

Peek 2019-09-30 10-49

@archmoj can you tell us why this happens? Note that, this mock is used on https://plot.ly/javascript/3d-scatter-plots/

**Comment by etpinard** [comment]
`gl3d_scatter3d_line3d_error3d_enable-alpha-with-rgba-color` also looks off:

image

Why are we blending the colors here? Shouldn't blending the opacity be sufficient?

**Comment by archmoj** [comment]
> Hmm. At first look, the new baseline for the `gl3d_opacity-scaling-spikes` mock looks worse than before:
> 
> Peek 2019-09-30 10-49
> 
> @archmoj can you tell us why this happens? Note that, this mock is used on https://plot.ly/javascript/3d-scatter-plots/

This happens because we need to blend colors and opacity values while the opacity values (as well as  the marker sizes) are too large on this mock.
On the mentioned mock if you reduce `marker.opacity` values (now at 0.8 and 0.9) or marker sizes values you would get a nice looking point cloud.
[codepen](https://codepen.io/MojtabaSamimi/pen/RwbzpdR?editors=0010)

**Comment by archmoj** [comment]
By applying opacity to our mock here are some of the cases where this PR can help fix:
Before (left) | After (right)
Screenshot from 2019-09-30 15-38-46
Screenshot from 2019-09-30 16-54-37
Screenshot from 2019-09-30 19-57-41
Screenshot from 2019-09-30 19-58-19
Screenshot from 2019-09-30 20-04-51
Screenshot from 2019-09-30 20-05-40
Screenshot from 2019-09-30 20-14-03

**Comment by archmoj** [comment]
> `gl3d_scatter3d_line3d_error3d_enable-alpha-with-rgba-color` also looks off:
> 
> image
> 
> Why are we blending the colors here? Shouldn't blending the opacity be sufficient?

Addressed in https://github.com/plotly/plotly.js/pull/4234/commits/40b1c6c7c3f1c911ea8e0a37fc06f5fcc409ac22.

**Comment by archmoj** [comment]
@etpinard would you mind review the changes after https://github.com/plotly/plotly.js/pull/4234/commits/40b1c6c7c3f1c911ea8e0a37fc06f5fcc409ac22?
It would be great to have it in 1.50.0
Thanks.

**Comment by etpinard** [comment]
@archmoj can you explain what you did in https://github.com/plotly/plotly.js/commit/40b1c6c7c3f1c911ea8e0a37fc06f5fcc409ac22 that changed the scatter baselines?

**Comment by etpinard** [comment]
Does this underlying mock:

image

have a set opacity? I can't say that either of the before and after rendering look good.

**Comment by etpinard** [comment]
This looks off to me:

image

Can you explain why the new baseline is suppositively better?

**Comment by archmoj** [comment]
> This looks off to me:
> 
> image
> 
> Can you explain why the new baseline is suppositively better?

Good call. That one is actually a `scatter3d` plot!
Fixed by https://github.com/plotly/plotly.js/pull/4234/commits/ebc99b896963d1fb61d09b3afa2b01cf1377cc9b.

**Comment by archmoj** [comment]
> Does this underlying mock:
> 
> image
> 
> have a set opacity? I can't say that either of the before and after rendering look good.

The opacity values are set to 0.5 for these two mocks.
The one on the left clearly has glitches. I could add another example in that regard.

**Comment by archmoj** [comment]
> @archmoj can you explain what you did in [40b1c6c](https://github.com/plotly/plotly.js/commit/40b1c6c7c3f1c911ea8e0a37fc06f5fcc409ac22) that changed the some `scatter3d` baselines?

For point-clouds (including `scatter3d` and `cone`) the depth mask is recorded when drawing new points.
Plus in respect to your https://github.com/plotly/plotly.js/pull/4234#issuecomment-536599104 for the mentioned traces alpha is applied (instead of color) for blending rgb values (i.e. similar to what used before).

**Comment by etpinard** [comment]
I'm curious @archmoj - what made you close this?

**Comment by archmoj** [comment]
Am working on another branch to make darker color blending not lighter.

**Pull request closed without merging**